File: benchmark/stress.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

from __future__ import print_function

import argparse
import collections
import contextlib
import functools
import itertools
import json
import math
import os
import random
import signal
import time

import redis
import pymysql


try:
    import httplib
except ImportError:
    import http.client as httplib

try:
    import urllib.parse as urllib
except ImportError:
    import urllib

from multiprocessing.pool import Pool, ThreadPool
from multiprocessing import Process

KEY_PREFIX = "stress_test:make_order"

USER_KEY = "{}:user".format(KEY_PREFIX)

SUCCESS_KEY = "{}:success".format(KEY_PREFIX)
FAILURE_KEY = "{}:failure".format(KEY_PREFIX)

ORDER_RESP_TIME_KEY = "{}:order_resp_time".format(KEY_PREFIX)
REQ_RESP_TIME_KEY = "{}:req_resp_time".format(KEY_PREFIX)

REQUEST_SUCCESS_KEY = "{}:request_success".format(KEY_PREFIX)
REQUEST_FAILURE_KEY = "{}:request_failure".format(KEY_PREFIX)

REQ_FINISH_TIME_KEY = "{}:req_finish_time".format(KEY_PREFIX)
ORDER_FINISH_TIME_KEY = "{}:order_finish_time".format(KEY_PREFIX)

redis_store = redis.Redis()
users, foods = {}, []


@contextlib.contextmanager
def db_query():
    db = pymysql.connect(host=os.getenv("DB_HOST", "localhost"),
                         port=int(os.getenv("DB_PORT", 3306)),
                         user=os.getenv("DB_USER", "root"),
                         passwd=os.getenv("DB_PASS", "toor"),
                         db=os.getenv("DB_NAME", "eleme"))
    try:
        yield db
    finally:
        db.close()


def load_users():
    global users
    with db_query() as db:
        cur = db.cursor()

        # load users
        cur.execute("SELECT id, name, password FROM user")

        for i, name, pw in cur.fetchall():
            users[i] = {"username": name, "password": pw}
    redis_store.sadd(USER_KEY, *users.keys())
    return users


def load_foods():
    global foods
    with db_query() as db:
        cur = db.cursor()
        cur.execute("SELECT id, stock, price FROM food")

        for i, stock, price in cur.fetchall():
            foods.append({"id": i, "stock": stock})
    return foods


def safe_loads(data):
    try:
        return json.loads(data)
    except:
        return data


class QueryException(Exception):

    def __init__(self, code, message):
        self.code = code
        self.message = message

    def __str__(self):
        return "{} {}".format(self.code, self.message)


class Query(object):

    __slots__ = ["access_token", "user_id", "cart_id", "client"]

    def __init__(self, host, port):
        self.client = httplib.HTTPConnection(host, port, timeout=3)

        self.access_token = None
        self.user_id = None
        self.cart_id = None

    def request(self, method, url, headers=None, data=None):
        data = data or {}
        headers = headers or {}
        headers["Content-Type"] = "application/json"

        start = time.time()
        status = None
        try:
            self.client.request(method, url, body=json.dumps(data),
                                headers=headers)
            response = self.client.getresponse()
            status = response.status
            data = response.read().decode("utf-8")
            self.client.close()
            return {"status": status, "data": safe_loads(data)}
        finally:
            now = time.time()
            elapsed = now - start

            with redis_store.pipeline() as p:
                if status in (200, 204):
                    p.incr(REQUEST_SUCCESS_KEY)
                    p.lpush(REQ_FINISH_TIME_KEY, now)
                else:
                    p.incr(REQUEST_FAILURE_KEY)
                p.lpush(REQ_RESP_TIME_KEY, elapsed)
                p.execute()

    def url(self, path):
        assert self.access_token
        params = {"access_token": self.access_token}
        qs = urllib.urlencode(params)
        return "{}?{}".format(path, qs) if qs else path

    def _do_login(self, username, password):
        data = {
            "username": username,
            "password": password
        }
        response = self.request("POST", "/login", data=data)
        if response["status"] == 200:
            self.access_token = response["data"]["access_token"]
            return True
        return False

    def login(self):
        user_id = redis_store.spop(USER_KEY)
        if not user_id:
            return False

        self.user_id = int(user_id)
        user = users[self.user_id]
        return self._do_login(user["username"], user["password"])

    def get_foods(self):
        res = self.request("GET", self.url("/foods"))
        return res["status"] == 200

    def get_orders(self):
        res = self.request("GET", self.url("/orders"))
        return res["status"] == 200

    def create_cart(self):
        response = self.request("POST", self.url("/carts"))
        try:
            self.cart_id = response["data"].get("cart_id")
        except:
            return False
        return response["status"] == 200

    def cart_add_food(self):
        food = random.choice(foods)
        data = {"food_id": food["id"], "count": 1}
        path = "/carts/{}".format(self.cart_id)
        res = self.request("PATCH", self.url(path), data=data)
        return res["status"] == 204

    def make_order(self):
        chain = [self.login, self.get_foods, self.create_cart,
                 self.cart_add_food, self.cart_add_food]
        for action in chain:
            if not action():
                return False

        data = {"cart_id": self.cart_id}
        res = self.request("POST", self.url("/orders"), data=data)
        return res["status"] == 200


def job(host, port):
    q = Query(host, port)

    start = time.time()
    try:
        ok = q.make_order()
    except:
        ok = False

    end = time.time()
    elapsed = end - start

    with redis_store.pipeline() as p:
        if ok:
            p.incr(SUCCESS_KEY)
            p.lpush(ORDER_FINISH_TIME_KEY, end)
        else:
            p.incr(FAILURE_KEY)
        p.lpush(ORDER_RESP_TIME_KEY, elapsed)
        p.execute()


def progress():
    try:
        prev = 0
        while True:
            time.sleep(1)
            cur = get_value(SUCCESS_KEY)

            msg = "Orders Per Second: {:4d}/s".format(cur - prev)
            print(msg, end='')
            print('\r' * len(msg), end='')

            prev = cur

    except KeyboardInterrupt:
        pass
    finally:
        print('\n')


def thread(host, port, threads, num):
    pool = ThreadPool(threads)
    for _ in range(num):
        pool.apply_async(job, (host, port))
        time.sleep(0.001)
    pool.close()
    pool.join()


def divide(n, m):
    """Divide integer n to m chunks
    """
    avg = int(n / m)
    remain = n - m * avg
    data = list(itertools.repeat(avg, m))
    for i in range(len(data)):
        if not remain:
            break
        data[i] += 1
        remain -= 1
    return data


def work(host, port, processes, threads, times):
    pool = Pool(processes,
                lambda: signal.signal(signal.SIGINT, signal.SIG_IGN))
    p = Process(target=progress)
    p.daemon = True

    start = time.time()

    try:
        for chunk in divide(times, processes):
            pool.apply_async(thread, (host, port, threads, chunk))

        p.start()

        pool.close()
        pool.join()
        p.terminate()
        p.join()

    except KeyboardInterrupt:
        pool.terminate()
        p.terminate()
        p.join()
        pool.join()

    return time.time() - start


def get_value(key):
    v = redis_store.get(key)
    return 0 if v is None else int(v)


def get_range(key):
    v = redis_store.lrange(key, 0, -1)
    return [float(i) for i in v]


def safe_div(a, b):
    return a / b if b else 0


def get_avg(l):
    return safe_div(sum(l), float(len(l)))


def report(processes, threads, total_time, total_order):
    success = get_value(SUCCESS_KEY)
    failure = get_value(FAILURE_KEY)
    req_success = get_value(REQUEST_SUCCESS_KEY)
    req_failure = get_value(REQUEST_FAILURE_KEY)

    req_resp_time = get_range(REQ_RESP_TIME_KEY)
    order_resp_time = get_range(ORDER_RESP_TIME_KEY)
    req_finish_time = get_range(REQ_FINISH_TIME_KEY)
    order_finish_time = get_range(ORDER_FINISH_TIME_KEY)

    assert len(order_resp_time) == success + failure
    assert len(req_resp_time) == req_success + req_failure

    req_avg = safe_div(sum(req_resp_time), float(req_success))
    order_avg = safe_div(sum(order_resp_time), success)

    req_sec = collections.Counter(int(t) for t in req_finish_time)
    order_sec = collections.Counter(int(t) for t in order_finish_time)

    # remove the highest and lowest score
    stats_req_sec = sorted(req_sec.values())[1:-1]
    max_req_sec = int(get_avg(stats_req_sec[-5:]))
    min_req_sec = int(get_avg(stats_req_sec[:5]))
    mean_req_sec = int(get_avg(stats_req_sec))

    # remove the highest and lowest score
    stats_order_sec = sorted(order_sec.values())[1:-1]
    max_order_sec = int(get_avg(stats_order_sec[-5:]))
    min_order_sec = int(get_avg(stats_order_sec[:5]))
    mean_order_sec = int(get_avg(stats_order_sec))

    p = functools.partial(print, sep='')

    p("Score:                ", max_order_sec)
    p("Correct Rate:         ", round(success / total_order * 100, 2), "%")

    p("\nStats")
    p("Concurrent Level:     ", processes, " x ", threads)
    p("Time taken for tests: ", round(total_time * 1000, 2), "ms")
    p("Complete requests:    ", req_success)
    p("Failed requests:      ", req_failure)
    p("Complete orders:      ", success)
    p("Failed orders:        ", failure)
    p("Time per request:     ", round(req_avg * 1000, 2), "ms", " (mean)")
    p("Time per order:       ", round(order_avg * 1000, 2), "ms", " (mean)")
    p("Request per second:   ", max_req_sec, " (max) ", min_req_sec, " (min) ", mean_req_sec, " (mean)")  # noqa
    p("Order per second:     ", max_order_sec, " (max) ", min_order_sec, " (min) ", mean_order_sec, " (mean)")  # noqa

    p("\nPercentage of orders made within a certain time (ms)")
    order_resp_time = sorted(set(order_resp_time)) if order_resp_time else [0]
    l = len(order_resp_time)
    for e in (0.5, 0.75, 0.8, 0.9, 0.95, 0.98, 1):
        idx = int(l * e)
        idx = 0 if idx == 0 else idx - 1
        p(" {:>4.0%}      ".format(e),
          int(math.ceil(order_resp_time[idx] * 1000)))


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument("-H", "--host", default="localhost",
                        help="server host name")
    parser.add_argument("-p", "--port", default=8080, type=int,
                        help="server port")
    parser.add_argument("-c", "--processes", default=2, type=int,
                        help="processes")
    parser.add_argument("-t", "--threads", default=4, type=int,
                        help="threads")
    parser.add_argument("-n", "--num", default=10000, type=int,
                        help="requests")

    args = parser.parse_args()

    redis_store.delete(
        USER_KEY, SUCCESS_KEY, FAILURE_KEY,
        ORDER_RESP_TIME_KEY, REQ_RESP_TIME_KEY,
        REQUEST_SUCCESS_KEY, REQUEST_FAILURE_KEY,
        REQ_FINISH_TIME_KEY, ORDER_FINISH_TIME_KEY)

    load_users()
    load_foods()

    total_time = work(
        args.host, args.port, args.processes, args.threads, args.num)

    report(args.processes, args.threads, total_time, float(args.num))


if __name__ == "__main__":
    main()

File: benchmark/stress.go
package main

import (
	"bytes"
	"database/sql"
	"encoding/json"
	"flag"
	"fmt"
	"io"
	"io/ioutil"
	"math/rand"
	"net/http"
	"net/url"
	"os"
	"runtime"
	"sort"
	"strings"
	"time"

	"github.com/garyburd/redigo/redis"
	_ "github.com/go-sql-driver/mysql"
)

//----------------------------------
// Stress Settings
//----------------------------------
const (
	MAX_COCURRENCY   = 10 * 1024
	REDIS_LOCAL_ADDR = "127.0.0.1:6379"
)

//----------------------------------
// Stress Abstracts
//----------------------------------
type Worker struct {
	r   *Reporter
	ctx struct {
		hosts []string
		port  int
	}
}

type Context struct {
	c      *http.Client
	w      *Worker
	user   User
	cartId string
}

type Reporter struct {
	orderMade       chan bool
	orderCost       chan time.Duration
	requestSent     chan bool
	userCurr        chan User
	numOrders       int
	cocurrency      int
	nOrderOk        int
	nOrderErr       int
	nOrderTotal     int
	nOrderPerSec    []int
	orderCosts      []time.Duration
	nRequestOk      int
	nRequestErr     int
	nRequestTotal   int
	nRequestPerSec  []int
	timeStampPerSec []int
	startAt         time.Time
	elapsed         time.Duration
}

//----------------------------------
// Entity Abstracts
//----------------------------------
type User struct {
	Id          int
	UserName    string
	PassWord    string
	AccessToken string
}

type Food struct {
	Id    int `json:"id"`
	Price int `json:"price"`
	Stock int `json:"stock"`
}

//----------------------------------
// Request JSON Bindings
//----------------------------------
type RequestLogin struct {
	UserName string `json:"username"`
	PassWord string `json:"password"`
}

type RequestCartAddFood struct {
	FoodId int `json:"food_id"`
	Count  int `json:"count"`
}

type RequestMakeOrder struct {
	CartId string `json:"cart_id"`
}

//----------------------------------
// Response JSON Bindings
//----------------------------------
type ResponseLogin struct {
	UserId      int    `json:"user_id"`
	UserName    string `json:"username"`
	AccessToken string `json:"access_token"`
}

type ResponseGetFoods []Food

type ResponseCreateCart struct {
	CartId string `json:"cart_id"`
}

type ResponseMakeOrder struct {
	Id   string `json:"id"`
	Item []struct {
		FoodId int
		Count  int
	} `json:"items"`
	Total int `json:"total"`
}

//----------------------------------
// Global Variables
//----------------------------------
var (
	users           = make([]User, 0)    // users
	foods           = make(map[int]Food) // map[food.Id]food
	isDebugMode     = false
	isReportToRedis = false
)

//----------------------------------
// Data Initialization
//----------------------------------
// Load all data.
func LoadData() {
	dbHost := os.Getenv("DB_HOST")
	dbPort := os.Getenv("DB_PORT")
	dbName := os.Getenv("DB_NAME")
	dbUser := os.Getenv("DB_USER")
	dbPass := os.Getenv("DB_PASS")

	if dbHost == "" {
		dbHost = "localhost"
	}
	if dbPort == "" {
		dbPort = "3306"
	}
	if dbName == "" {
		dbName = "eleme"
	}
	if dbUser == "" {
		dbUser = "root"
	}
	if dbPass == "" {
		dbPass = "toor"
	}

	fmt.Printf("Connect to mysql..")
	dbDsn := fmt.Sprintf("%s:%s@tcp(%s:%s)/%s", dbUser, dbPass, dbHost, dbPort, dbName)
	db, err := sql.Open("mysql", dbDsn)
	if err != nil {
		panic(err)
	}
	defer db.Close()
	fmt.Printf("OK\n")
	fmt.Printf("Ping to mysql..")
	err = db.Ping()
	if err != nil {
		panic(err)
	}
	fmt.Printf("OK\n")
	fmt.Printf("Load users from mysql..")
	LoadUsers(db)
	fmt.Printf("OK\n")
	fmt.Printf("Load foods from mysql..")
	LoadFoods(db)
	fmt.Printf("OK\n")
}

// Load users from mysql
func LoadUsers(db *sql.DB) {
	var user User
	rows, err := db.Query("SELECT `id`, `name`, `password` from user")
	if err != nil {
		panic(err)
	}
	defer rows.Close()
	for rows.Next() {
		err = rows.Scan(&user.Id, &user.UserName, &user.PassWord)
		if err != nil {
			panic(err)
		}
		users = append(users, user)
	}
}

// Load foods from mysql
func LoadFoods(db *sql.DB) {
	var food Food
	rows, err := db.Query("SELECT `id`, `stock`, `price` from food")
	if err != nil {
		panic(err)
	}
	defer rows.Close()
	for rows.Next() {
		err = rows.Scan(&food.Id, &food.Stock, &food.Price)
		if err != nil {
			panic(err)
		}
		foods[food.Id] = food
	}
}

//----------------------------------
// Request Utils
//----------------------------------
// Build url with path and parameters.
func (w *Worker) Url(path string, params url.Values) string {
	// random choice one host for load balance
	i := rand.Intn(len(w.ctx.hosts))
	host := w.ctx.hosts[i]
	s := fmt.Sprintf("http://%s:%d%s", host, w.ctx.port, path)
	if params == nil {
		return s
	}
	p := params.Encode()
	return fmt.Sprintf("%s?%s", s, p)
}

// Get json from uri.
func (w *Worker) Get(c *http.Client, url string, bind interface{}) (int, error) {
	r, err := c.Get(url)
	if err != nil {
		if r != nil {
			ioutil.ReadAll(r.Body)
			r.Body.Close()
		}
		return 0, err
	}
	defer r.Body.Close()
	err = json.NewDecoder(r.Body).Decode(bind)
	if bind == nil {
		return r.StatusCode, nil
	}
	return r.StatusCode, err
}

// Post json to uri and get json response.
func (w *Worker) Post(c *http.Client, url string, data interface{}, bind interface{}) (int, error) {
	var body io.Reader
	if data != nil {
		bs, err := json.Marshal(data)
		if err != nil {
			return 0, err
		}
		body = bytes.NewReader(bs)
	}
	r, err := c.Post(url, "application/json", body)
	if err != nil {
		if r != nil {
			ioutil.ReadAll(r.Body)
			r.Body.Close()
		}
		return 0, err
	}
	defer r.Body.Close()
	err = json.NewDecoder(r.Body).Decode(bind)
	if bind == nil {
		return r.StatusCode, nil
	}
	return r.StatusCode, err
}

// Patch url with json.
func (w *Worker) Patch(c *http.Client, url string, data interface{}, bind interface{}) (int, error) {
	bs, err := json.Marshal(data)
	if err != nil {
		return 0, err
	}
	req, err := http.NewRequest("PATCH", url, bytes.NewReader(bs))
	if err != nil {
		return 0, err
	}
	req.Header.Set("Content-Type", "application/json")
	res, err := c.Do(req)
	if err != nil {
		if res != nil {
			ioutil.ReadAll(res.Body)
			res.Body.Close()
		}
		return 0, err
	}
	defer res.Body.Close()
	err = json.NewDecoder(res.Body).Decode(bind)
	if res.StatusCode == http.StatusNoContent || bind == nil {
		return res.StatusCode, nil
	}
	return res.StatusCode, err
}

//----------------------------------
//  Order Handle Utils
//----------------------------------
// Random choice a food. Dont TUCAO this function,
// it works and best O(1).
func GetRandFood() Food {
	for {
		idx := rand.Intn(len(foods))
		food, ok := foods[idx+1]
		if ok {
			return food
		}
	}
}

//----------------------------------
//  Work Job Context
//----------------------------------
func (ctx *Context) UrlWithToken(path string) string {
	user := ctx.user
	params := url.Values{}
	params.Add("access_token", user.AccessToken)
	return ctx.w.Url(path, params)
}

func (ctx *Context) Login() bool {
	user := ctx.user
	data := &RequestLogin{user.UserName, user.PassWord}
	body := &ResponseLogin{}
	url := ctx.w.Url("/login", nil)
	statusCode, err := ctx.w.Post(ctx.c, url, data, body)
	if err != nil {
		if isDebugMode {
			fmt.Printf("Request login error: %v\n", err)
		}
		ctx.w.r.requestSent <- false
		return false
	}
	if statusCode == http.StatusOK {
		ctx.user.AccessToken = body.AccessToken
		ctx.w.r.requestSent <- true
		return true
	}
	ctx.w.r.requestSent <- false
	return false
}

func (ctx *Context) GetFoods() bool {
	// body := &ResponseGetFoods{}
	url := ctx.UrlWithToken("/foods")
	statusCode, err := ctx.w.Get(ctx.c, url, nil)
	if err != nil {
		if isDebugMode {
			fmt.Printf("Request get foods error: %v\n", err)
		}
		ctx.w.r.requestSent <- false
		return false
	}
	if statusCode == http.StatusOK {
		ctx.w.r.requestSent <- true
		return true
	}
	ctx.w.r.requestSent <- false
	return false
}

func (ctx *Context) CreateCart() bool {
	body := &ResponseCreateCart{}
	url := ctx.UrlWithToken("/carts")
	statusCode, err := ctx.w.Post(ctx.c, url, nil, body)
	if err != nil {
		if isDebugMode {
			fmt.Printf("Request create carts error: %v\n", err)
		}
		ctx.w.r.requestSent <- false
		return false
	}
	if statusCode == http.StatusOK {
		ctx.cartId = body.CartId
		ctx.w.r.requestSent <- true
		return true
	}
	ctx.w.r.requestSent <- false
	return false
}

func (ctx *Context) CartAddFood() bool {
	path := fmt.Sprintf("/carts/%s", ctx.cartId)
	url := ctx.UrlWithToken(path)
	food := GetRandFood()
	data := &RequestCartAddFood{food.Id, 1}
	statusCode, err := ctx.w.Patch(ctx.c, url, data, nil)
	if err != nil {
		if isDebugMode {
			fmt.Printf("Request error cart add food error: %v\n", err)
		}
		ctx.w.r.requestSent <- false
		return false
	}
	if statusCode == http.StatusNoContent {
		ctx.w.r.requestSent <- true
		return true
	}
	ctx.w.r.requestSent <- false
	return false
}

func (ctx *Context) MakeOrder() bool {
	if !ctx.Login() || !ctx.GetFoods() || !ctx.CreateCart() {
		return false
	}
	count := rand.Intn(3) + 1
	for i := 0; i < count; i++ {
		if !ctx.CartAddFood() {
			return false
		}
	}
	data := &RequestMakeOrder{ctx.cartId}
	body := &ResponseMakeOrder{}
	url := ctx.UrlWithToken("/orders")
	statusCode, err := ctx.w.Post(ctx.c, url, data, body)
	if err != nil {
		if isDebugMode {
			fmt.Printf("Request make order error: %v\n", err)
		}
		ctx.w.r.requestSent <- false
		return false
	}
	if statusCode == http.StatusOK {
		ctx.w.r.requestSent <- true
		return true
	}
	ctx.w.r.requestSent <- false
	return false
}

//----------------------------------
// Worker
//----------------------------------
func NewWorker(hosts []string, port int, r *Reporter) *Worker {
	w := &Worker{}
	w.r = r
	w.ctx.hosts = hosts
	w.ctx.port = port
	return w
}
func (w *Worker) Work() {
	ctx := &Context{}
	ctx.w = w
	t := &http.Transport{}
	ctx.c = &http.Client{
		Timeout:   3 * time.Second,
		Transport: t,
	}
	for {
		t.CloseIdleConnections()
		startAt := time.Now()
		ctx.user = <-w.r.userCurr
		w.r.orderMade <- ctx.MakeOrder()
		w.r.orderCost <- time.Since(startAt)
	}
}

//----------------------------------
// Statstics Reporter
//----------------------------------
// Create reporter
func NewReporter(numOrders int, cocurrency int) *Reporter {
	return &Reporter{
		make(chan bool, cocurrency),
		make(chan time.Duration, cocurrency),
		make(chan bool, cocurrency),
		make(chan User, cocurrency),
		numOrders,
		cocurrency,
		0,
		0,
		0,
		make([]int, 0),
		make([]time.Duration, 0),
		0,
		0,
		0,
		make([]int, 0),
		make([]int, 0),
		time.Now(),
		0,
	}
}

// Start reporter
func (r *Reporter) Start() {
	r.startAt = time.Now()
	go func() {
		t := time.NewTicker(1 * time.Second)
		for {
			nOrderOk := r.nOrderOk
			nRequestOk := r.nRequestOk
			<-t.C
			nOrderPerSec := r.nOrderOk - nOrderOk
			r.nOrderPerSec = append(r.nOrderPerSec, nOrderPerSec)
			nRequestPerSec := r.nRequestOk - nRequestOk
			r.nRequestPerSec = append(r.nRequestPerSec, nRequestPerSec)
			r.timeStampPerSec = append(r.timeStampPerSec, time.Now().Second())
			fmt.Printf("Finished orders: %d\n", nOrderPerSec)
		}
	}()
	go func() {
		for {
			orderMade := <-r.orderMade
			orderCost := <-r.orderCost
			if orderMade {
				r.nOrderOk = r.nOrderOk + 1
				r.orderCosts = append(r.orderCosts, orderCost)
			} else {
				r.nOrderErr = r.nOrderErr + 1
			}
			r.nOrderTotal = r.nOrderTotal + 1
			if r.nOrderTotal >= r.numOrders {
				r.Stop()
			}
		}
	}()
	go func() {
		for {
			requestSent := <-r.requestSent
			if requestSent {
				r.nRequestOk = r.nRequestOk + 1
			} else {
				r.nRequestErr = r.nRequestErr + 1
			}
			r.nRequestTotal = r.nRequestTotal + 1
		}
	}()
	for i := 0; i < len(users); i++ {
		r.userCurr <- users[i]
	}
	timeout := time.After(3 * time.Second)
	for r.nOrderTotal < r.numOrders {
		select {
		case <-timeout:
			r.Stop()
		}
	}
	r.Stop()
}

// Stop the reporter and exit full process.
func (r *Reporter) Stop() {
	r.elapsed = time.Since(r.startAt)
	r.Report()
	os.Exit(0)
}

// Report stats to console and redis.
func (r *Reporter) Report() {
	//---------------------------------------------------
	// Report to console
	//---------------------------------------------------
	sort.Ints(r.nOrderPerSec)
	sort.Ints(r.nRequestPerSec)
	nOrderPerSecMax := MeanOfMaxFive(r.nOrderPerSec)
	nOrderPerSecMin := MeanOfMinFive(r.nOrderPerSec)
	nOrderPerSecMean := Mean(r.nOrderPerSec)
	nRequestPerSecMax := MeanOfMaxFive(r.nRequestPerSec)
	nRequestPerSecMin := MeanOfMinFive(r.nRequestPerSec)
	nRequestPerSecMean := Mean(r.nRequestPerSec)
	sort.Ints(r.nRequestPerSec)
	orderCostSeconds := []float64{}
	for i := 0; i < len(r.orderCosts); i++ {
		orderCostSeconds = append(orderCostSeconds, r.orderCosts[i].Seconds())
	}
	sort.Float64s(orderCostSeconds)
	msTakenTotal := int(r.elapsed.Nanoseconds() / 1000000.0)
	msPerOrder := MeanFloat64(orderCostSeconds) * 1000.0
	msPerRequest := SumFloat64(orderCostSeconds) * 1000.0 / float64(r.nRequestOk)
	//---------------------------------------------------
	// Report to console
	//---------------------------------------------------
	fmt.Print("\nStats\n")
	fmt.Printf("Concurrency level:         %d\n", r.cocurrency)
	fmt.Printf("Time taken for tests:      %dms\n", msTakenTotal)
	fmt.Printf("Complete requests:         %d\n", r.nRequestOk)
	fmt.Printf("Failed requests:           %d\n", r.nRequestErr)
	fmt.Printf("Complete orders:           %d\n", r.nOrderOk)
	fmt.Printf("Failed orders:             %d\n", r.nOrderErr)
	fmt.Printf("Time per request:          %.2fms\n", msPerRequest)
	fmt.Printf("Time per order:            %.2fms\n", msPerOrder)
	fmt.Printf("Request per second:        %d (max)  %d (min)  %d(mean)\n", nRequestPerSecMax, nRequestPerSecMin, nRequestPerSecMean)
	fmt.Printf("Order per second:          %d (max)  %d (min)  %d (mean)\n\n", nOrderPerSecMax, nOrderPerSecMin, nOrderPerSecMean)
	fmt.Printf("Percentage of orders made within a certain time (ms)\n")
	if len(orderCostSeconds) == 0 {
		return
	}
	percentages := []int{50, 75, 80, 90, 95, 98, 100}
	for _, percentage := range percentages {
		idx := int(float64(percentage*len(orderCostSeconds)) / float64(100.0))
		if idx > 0 {
			idx = idx - 1
		} else {
			idx = 0
		}
		orderCostSecond := orderCostSeconds[idx]
		fmt.Printf("%d%%\t%d\n", percentage, int(orderCostSecond*1000.0))
	}
	//---------------------------------------------------
	// Report to redis
	//---------------------------------------------------
	if !isReportToRedis {
		return
	}
	conn, err := redis.Dial("tcp", REDIS_LOCAL_ADDR)
	defer conn.Close()
	if err != nil {
		panic(err)
	}
	conn.Do("SET", "stress_test:make_order:success", r.nOrderOk)
	conn.Do("SET", "stress_test:make_order:failure", r.nOrderErr)
	conn.Do("SET", "stress_test:make_order:req_success", r.nRequestErr)
	conn.Do("SET", "stress_test:make_order:req_failure", r.nRequestErr)
	conn.Do("SET", "stress_test:make_order:max_order_sec", nOrderPerSecMax)
	conn.Do("SET", "stress_test:make_order:min_order_sec", nOrderPerSecMin)
	conn.Do("SET", "stress_test:make_order:mean_order_sec", nOrderPerSecMean)
	conn.Do("SET", "stress_test:make_order:max_req_sec", nRequestPerSecMax)
	conn.Do("SET", "stress_test:make_order:min_req_sec", nRequestPerSecMin)
	conn.Do("SET", "stress_test:make_order:mean_req_sec", nRequestPerSecMean)
	conn.Do("SET", "stress_test:make_order:time_per_order", msPerOrder)
	conn.Do("SET", "stress_test:make_order:time_per_req", msPerRequest)
	for i := 0; i < len(r.timeStampPerSec); i++ {
		conn.Do("HSET", "stress_test:make_order:order_sec", r.timeStampPerSec[i], r.nOrderPerSec[i])
		conn.Do("HSET", "stress_test:make_order:req_sec", r.timeStampPerSec[i], r.nRequestPerSec[i])
	}
}

//----------------------------------
// Math util functions
//----------------------------------
func MeanOfMaxFive(sortedArr []int) int {
	if len(sortedArr) == 0 {
		return 0
	}
	if len(sortedArr) == 1 {
		return sortedArr[0]
	}
	if len(sortedArr) == 2 {
		return sortedArr[1]
    }
	sortedArr = sortedArr[1 : len(sortedArr)-1]
	if len(sortedArr) > 5 {
		return Mean(sortedArr[len(sortedArr)-5:])
	}
	return sortedArr[len(sortedArr)-1]
}

func MeanOfMinFive(sortedArr []int) int {
	if len(sortedArr) == 0 {
		return 0
	}
	if len(sortedArr) == 1 {
		return sortedArr[0]
	}
	if len(sortedArr) == 2 {
		return sortedArr[0]
    }
	sortedArr = sortedArr[1 : len(sortedArr)-1]
	if len(sortedArr) > 5 {
		return Mean(sortedArr[0:5])
	}
	return sortedArr[0]
}

func Mean(arr []int) int {
	if len(arr) == 0 {
		return 0
	}
	sum := 0
	for i := 0; i < len(arr); i++ {
		sum = sum + arr[i]
	}
	return int(float64(sum) / float64(len(arr)))
}

func MeanFloat64(arr []float64) float64 {
	return SumFloat64(arr) / float64(len(arr))
}

func SumFloat64(arr []float64) float64 {
	if len(arr) == 0 {
		return 0
	}
	sum := 0.0
	for i := 0; i < len(arr); i++ {
		sum = sum + arr[i]
	}
	return sum
}

//----------------------------------
// Main
//----------------------------------
func main() {
	runtime.GOMAXPROCS(runtime.NumCPU())
	//----------------------------------
	// Arguments parsing and validation
	//----------------------------------
	hosts := flag.String("h", "localhost", "server hosts, split by comma")
	port := flag.Int("p", 8080, "server port")
	cocurrency := flag.Int("c", 1000, "request cocurrency")
	numOrders := flag.Int("n", 1000, "number of orders to perform")
	debug := flag.Bool("d", false, "debug mode")
	reportRedis := flag.Bool("r", true, "report to local redis")
	flag.Parse()
	if flag.NFlag() == 0 {
		flag.PrintDefaults()
		os.Exit(1)
	}
	if *debug {
		isDebugMode = true
	}
	if *reportRedis {
		isReportToRedis = true
	}
	//----------------------------------
	// Validate cocurrency
	//----------------------------------
	if *cocurrency > MAX_COCURRENCY {
		fmt.Printf("Exceed max cocurrency (is %d)", MAX_COCURRENCY)
		os.Exit(1)
	}
	//----------------------------------
	// Load users/foods and work
	//----------------------------------
	LoadData()
	reporter := NewReporter(*numOrders, *cocurrency)
	for i := 0; i < *cocurrency; i++ {
		go func() {
			w := NewWorker(strings.Split(*hosts, ","), *port, reporter)
			w.Work()
		}()
	}
	// start reporter
	reporter.Start()
}
